### PR TITLE
[MRG] Faster test

### PR DESCRIPTION
**Pull request by nouiz**
Do not merge now, more to come. I want travis to test one of the change.

NEWS.txt:
- This PR change how we handle mode optimizer. If it is a query object, we cache the query the first time we need it. This allow to speed up tests. But this change the interface. If some optimization get registered after, the mode won't requery it, so it won't find the new registered optimization.

### CODE REVIEW

**Comment by nouiz** [comment]
@abergeron @lamblin there is an interface change in the first commit of this PR. What do you think of this?

File: theano/compile/mode.py [hunk 1]
```diff
@@ -290,7 +290,10 @@ def __str__(self):
 
     def __get_optimizer(self):
         if isinstance(self._optimizer, gof.Query):
-            return optdb.query(self._optimizer)
+            if hasattr(self, '_optimizer_cache'):
+                return self._optimizer_cache
+            self._optimizer_cache = optdb.query(self._optimizer)
+            return self._optimizer_cache
```
Comment (lamblin) [inline theano/compile/mode.py]: Are there any config flags that could change during the execution, and that could have an impact on that?
If so, we could just prevent it from being modified during the execution.

Comment (nouiz) [inline theano/compile/mode.py]: I agree this need more taught. I'll remove it from this PR.

**Comment by nouiz** [comment]
I removed the controversial part. We found other stuff to optimize. I'll work on them this week in another PR. When travis pass, I think we can merge after review.

File: theano/gof/graph.py [hunk 2]
```diff
@@ -786,26 +794,54 @@ def io_toposort(inputs, outputs, orderings=None):
...
-                rval = list(obj.inputs)
-            rval.extend(orderings.get(obj, []))
-        else:
-            assert not orderings.get(obj, [])
-        return rval
+    # We build 2 functions as a speed up
+    deps_cache = {}
+
+    deps = None
+    _deps = None
```
Comment (lamblin) [inline theano/gof/graph.py]: Rename _deps

File: theano/gof/graph.py [hunk 3]
```diff
@@ -727,22 +728,29 @@ def general_toposort(r_out, deps, debug_print=False):
...
 
-    def _deps(io):
-        if io not in deps_cache:
-            d = deps(io)
-            if d:
-                if not isinstance(d, (list, OrderedSet)):
-                    raise TypeError("Non-deterministic collections here make"
+    :note: deps should be provided or can be None and the caller
+        provide _deps and deps_cache. The second option remove a
+        Python function call, so is faster.
```
Comment (lamblin) [inline theano/gof/graph.py]: Update doc to specify what deps, _deps and deps_cache should be (their type)

File: theano/tensor/tests/test_opt.py [hunk 4]
```diff
@@ -644,7 +640,7 @@ def test_multiple_case(self):
                 assert numpy.allclose(out, numpy.sign(val_inputs[0]) * 2 / 3)
                 assert(out_dtype == out.dtype)
         finally:
-            mode._optimizer = old_optimizer
+            pass
```
Comment (lamblin) [inline theano/tensor/tests/test_opt.py]: remove the try/finally if it is not used

File: theano/tensor/tests/test_opt.py [hunk 5]
```diff
@@ -749,7 +744,7 @@ def test_multiple_case_that_fail(self):
                 assert(out_dtype == out.dtype)
 
         finally:
-            mode._optimizer = old_optimizer
+            pass
```
Comment (lamblin) [inline theano/tensor/tests/test_opt.py]: Idem

**Comment by nouiz** [comment]
I merged it by mistake. The small modif was cosmetic, I have done them in this PR:

gh-2823

**Pull request merged**